<a href="https://colab.research.google.com/github/cskipper07/Data-Science/blob/main/2_EDA_cnm_copy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Exploratory Data Analysis: CNM**
* Code for renaming, collapsing sides, and recoding some cnm datasets is in the A2_cnm.ipynb (Aim 2 cranial nonmetrics) file.
* Some outputs have been removed to protect PII and the raw data.

### Input data file:
*   *cnm_cleaned.xlsx*: This is the output file from the (Intra)Observer Error folder with: 1) renamed traits using codes and 2) traits removed with low IO agreement. I added a blank row with all 9s at the end of this cleaned file to fix the decimal issue. In this document, I replaced all 9s with np.nan. Then I removed the blank row (index 97).
*   *trait codes and names_unsided.xlsx*: This file contains the unsided trait codes and names for the cnm data.


### Import libraries

In [ ]:
pip install dtale

In [ ]:
import dtale

In [ ]:
# Import libraries
import pandas as pd
import scipy
import numpy as np
import seaborn as sns
#import dtale
import os

In [ ]:
!pip install --upgrade openpyxl

### Set print options

In [ ]:
import sys

In [ ]:
import pandas as pd
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)
pd.set_option('display.width', 1000)

In [ ]:
np.set_printoptions(threshold=sys.maxsize)

### Set export

In [ ]:
from google.colab import  drive
drive.mount('/drive')

Mounted at /drive


### Import data

In [ ]:
cnm = pd.read_excel('cnm_cleaned.xlsx')

In [ ]:
cnm.replace(9, np.nan, inplace=True)

In [ ]:
cnm.head()

In [ ]:
cnm.tail()

In [ ]:
cnm = cnm.drop(97)

In [ ]:
cnm.tail()

In [ ]:
import dtale.app as dtale_app

dtale_app.USE_COLAB = True

In [ ]:
dtale.show(cnm)

In [ ]:
cnm.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 97 entries, 0 to 96
Data columns (total 85 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   SkelID       97 non-null     object 
 1   Collection   97 non-null     object 
 2   Sex          97 non-null     object 
 3   Age          97 non-null     float64
 4   Population   97 non-null     object 
 5   Population2  19 non-null     object 
 6   Population3  10 non-null     object 
 7   Population4  1 non-null      object 
 8   SONR         96 non-null     float64
 9   SONL         97 non-null     float64
 10  SOFR         96 non-null     float64
 11  SOFL         97 non-null     float64
 12  IFSR         94 non-null     float64
 13  IFSL         94 non-null     float64
 14  MIFR         96 non-null     float64
 15  ZFFR         97 non-null     float64
 16  ZFFL         97 non-null     float64
 17  CCOR         97 non-null     float64
 18  CCOL         96 non-null     float64
 19  SSSF      

## Frequency counts

In [ ]:
cnm['SONR'].value_counts(normalize=True)

1.0    0.62500
0.0    0.25000
2.0    0.09375
4.0    0.03125
Name: SONR, dtype: float64

In [ ]:
cnm['SONR'].value_counts()

1.0    60
0.0    24
2.0     9
4.0     3
Name: SONR, dtype: int64

In [ ]:
# calculate number of 0 scores in SONR
cnm[cnm['SONR'] == 0].shape[0]

24

In [ ]:
# cnm[cnm['SONR'].values == 0].count()

In [ ]:
# calculate number of non-null (non-NA) scores in SONR
cnm['SONR'].notnull().sum()


96

In [ ]:
print('Frequency of score=0:', '\t', (cnm[cnm['SONR'] == 0].shape[0]) / (cnm['SONR'].notnull().sum()))
print('Frequency of score=1:', '\t', (cnm[cnm['SONR'] == 1].shape[0]) / (cnm['SONR'].notnull().sum()))
print('Frequency of score=2:', '\t', (cnm[cnm['SONR'] == 2].shape[0]) / (cnm['SONR'].notnull().sum()))
print('Frequency of score=3:', '\t', (cnm[cnm['SONR'] == 3].shape[0]) / (cnm['SONR'].notnull().sum()))
print('Frequency of score=4:', '\t', (cnm[cnm['SONR'] == 4].shape[0]) / (cnm['SONR'].notnull().sum()))
print('Frequency of score=5:', '\t', (cnm[cnm['SONR'] == 5].shape[0]) / (cnm['SONR'].notnull().sum()))

Frequency of score=0: 	 0.25
Frequency of score=1: 	 0.625
Frequency of score=2: 	 0.09375
Frequency of score=3: 	 0.0
Frequency of score=4: 	 0.03125
Frequency of score=5: 	 0.0


In [ ]:
# slice from column 8 because that is the first cnm trait column

cnm_VarsOnly = pd.DataFrame(cnm.iloc[:, 8:])

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SONR, SONL, SOFR, SOFL....

In [ ]:
cnm_VarsOnly.head()

In [ ]:
column_names = ('Trait Code', 0, 1, 2, 3, 4, 5)
cnm_freq_counts = pd.DataFrame(columns = column_names)
for i in cnm_VarsOnly.columns:
    #print(cnm[i].value_counts(normalize=True))
    freq_0 = (cnm_VarsOnly[cnm_VarsOnly[i] == 0].shape[0]) / (cnm_VarsOnly[i].notnull().sum())
    freq_1 = (cnm_VarsOnly[cnm_VarsOnly[i] == 1].shape[0]) / (cnm_VarsOnly[i].notnull().sum())
    freq_2 = (cnm_VarsOnly[cnm_VarsOnly[i] == 2].shape[0]) / (cnm_VarsOnly[i].notnull().sum())
    freq_3 = (cnm_VarsOnly[cnm_VarsOnly[i] == 3].shape[0]) / (cnm_VarsOnly[i].notnull().sum())
    freq_4 = (cnm_VarsOnly[cnm_VarsOnly[i] == 4].shape[0]) / (cnm_VarsOnly[i].notnull().sum())
    freq_5 = (cnm_VarsOnly[cnm_VarsOnly[i] == 5].shape[0]) / (cnm_VarsOnly[i].notnull().sum())
    #freq_6 = (cnm_VarsOnly[cnm_VarsOnly[i] == 6].shape[0]) / (cnm_VarsOnly[i].notnull().sum())
    print(i,'\t', freq_0,'\t', freq_1,'\t', freq_2,'\t', freq_3,'\t', freq_4,'\t', freq_5)
    cnm_freq_counts.loc[i] = [i, round(freq_0, 2), round(freq_1, 2), round(freq_2, 2), round(freq_3, 2), round(freq_4, 2), round(freq_5, 2)]

SONR 	 0.25 	 0.625 	 0.09375 	 0.0 	 0.03125 	 0.0
SONL 	 0.25773195876288657 	 0.6494845360824743 	 0.08247422680412371 	 0.0 	 0.010309278350515464 	 0.0
SOFR 	 0.4479166666666667 	 0.3854166666666667 	 0.16666666666666666 	 0.0 	 0.0 	 0.0
SOFL 	 0.5154639175257731 	 0.3711340206185567 	 0.1134020618556701 	 0.0 	 0.0 	 0.0
IFSR 	 0.3829787234042553 	 0.20212765957446807 	 0.4148936170212766 	 0.0 	 0.0 	 0.0
IFSL 	 0.3191489361702128 	 0.32978723404255317 	 0.35106382978723405 	 0.0 	 0.0 	 0.0
MIFR 	 0.6354166666666666 	 0.15625 	 0.17708333333333334 	 0.03125 	 0.0 	 0.0
ZFFR 	 0.041237113402061855 	 0.32989690721649484 	 0.6288659793814433 	 0.0 	 0.0 	 0.0
ZFFL 	 0.030927835051546393 	 0.31958762886597936 	 0.6391752577319587 	 0.010309278350515464 	 0.0 	 0.0
CCOR 	 0.31958762886597936 	 0.6804123711340206 	 0.0 	 0.0 	 0.0 	 0.0
CCOL 	 0.3958333333333333 	 0.6041666666666666 	 0.0 	 0.0 	 0.0 	 0.0
SSSF 	 0.010416666666666666 	 0.8333333333333334 	 0.07291666666666667 	 0.08

In [ ]:
cnm_freq_counts.tail()

Trait Code     0     1     2    3    4    5
PALT       PALT  0.20  0.80  0.00  0.0  0.0  0.0
APIC       APIC  0.76  0.24  0.00  0.0  0.0  0.0
METO       METO  0.99  0.00  0.01  0.0  0.0  0.0
MANT       MANT  0.62  0.38  0.00  0.0  0.0  0.0
PHAR       PHAR  0.75  0.25  0.00  0.0  0.0  0.0

In [ ]:
cnm_freq_counts.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_frequency_counts.xlsx', index=False)

In [ ]:
'''
for i in cnm.columns:
  print(cnm[i].value_counts(normalize=True))
  '''

## Collapse Left/Right

In [ ]:
'''
selected_columns = cnm[['SkelID', 'Collection', 'ANS', 'INA', 'IOB', 'NAS', 'NAW', 'NBC', 'NBS', 'NFS', 'OBS', 'TPS', ]]
Index into desired columns to copy

cnm_collapsed = selected_columns.copy()
'''

Traits removed due to low IO agreement:
*   cnm['SAGB']
*   cnm['MFLoL']
*   cnm['LBSO']
*   cnm['MIFL']
*   cnm['SPS']
*   cnm['OMBR']
*   cnm['PBD']
*   cnm['CIVL']
*   cnm['PTBR']
*   cnm['PS']
*   cnm['MVP']
*   cnm['TWIG']

##### cnm_collapsing

In [ ]:
cnm_collapsing = pd.DataFrame(cnm_VarsOnly)

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SONR, SONL, SOFR, SOFL....

In [ ]:
cnm_collapsing.head()

In [ ]:
# make a list of the traits to retain. These traits have had their opposite side removed due to low IO agreement

retained_traits = set(['MFLoR', 'MIFR', 'OMBL', 'CIVR', 'PTBL'])

In [ ]:
our_column_names = []                    # create an empty list for the column names
for t in list(cnm_collapsing.columns):
    if t[-1] == 'L':                     # find all column names that end in L (double check that on Lefts end in L prior to this step)
        our_column_names.append(t[:-1])  # t[:-1] truncates the last letter (L or R) from the col name
print(our_column_names)

['SON', 'SOF', 'IFS', 'ZFF', 'CCO', 'FOI', 'FSI', 'PTB', 'TYM', 'AUD', 'MT', 'NO', 'PZT', 'ZS', 'LBM', 'LBLa', 'PF', 'MF', 'CRB', 'EPB', 'FTA', 'PNB', 'AST', 'OMB', 'HYP', 'APF', 'FF', 'MHB', 'MEN']


The output for the following code cell was removed to protect PII. The output contained the following columns:
* SONR, SONL, SOFR, SOFL....

In [ ]:
cnm_collapsing.head()

The output for the following code cell was removed to protect PII.

In [ ]:
for c in our_column_names:
    c_l = c +'L'                                # loop through all Lefts in variable c_l
    c_r = c + 'R'                               # loop through all Rights in variable c_r
    #print(value)
    cnm_collapsing[c] = cnm_collapsing[c_l]                     # overwrite the base (unsided) column variables (c) with Left values (c_l)
    cnm_collapsing['mask'] = (cnm_collapsing[c].isnull())       # create a mask variable where all NaNs will be True
    if c_l in retained_traits or c_r in retained_traits:
        if c_r in retained_traits:
          cnm_collapsing[c] = cnm_collapsing[c_r]
          continue
        if c_l in retained_traits:
          continue
    cnm_collapsing.loc[cnm_collapsing['mask'], c] = cnm_collapsing[c_r] # for rows that have NaNs in 'mask', overwrite the corresponding value in c with the value from the Right side (c_r)
    print(cnm_collapsing[c].tail(15))                   # print the unsided columns tail to check for
#print(cnm_collapsing['SON'].head())
#print(cnm_collapsing['FOI'].tail(10))

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SONR, SONL, SOFR, SOFL....
* included a 'mask' column

In [ ]:
cnm_collapsing.head()

In [ ]:
del cnm_collapsing['mask']

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SONR, SONL, SOFR, SOFL....
* 'mask' column removed

In [ ]:
cnm_collapsing.head()

In [ ]:
cnm_collapsing.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 97 entries, 0 to 96
Columns: 106 entries, SONR to MEN
dtypes: float64(106)
memory usage: 83.1 KB


##### cnm_collapsing_unsided

In [ ]:
cnm_collapsing_unsided = pd.DataFrame(cnm_collapsing.loc[:, 'SON':])

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SON, SOF, IFS....

In [ ]:
cnm_collapsing_unsided.head()

##### cnm_unsided

In [ ]:
# collected the unsided names from the original cnm df

unsided_names = []
for c in cnm.columns:
  if c[-1] == 'L' or c[-1] == 'R':
    continue
  unsided_names.append(c)
print(unsided_names)

['SkelID', 'Collection', 'Sex', 'Age', 'Population', 'Population2', 'Population3', 'Population4', 'SSSF', 'ANS', 'INA', 'IOB', 'NAS', 'NAW', 'NBC', 'NBS', 'NFS', 'OBS', 'TPS', 'INCA', 'BREG', 'PALT', 'APIC', 'METO', 'MANT']


In [ ]:
cnm_unsided = pd.DataFrame()

In [ ]:
for c in unsided_names:
  cnm_unsided[c] = cnm[c]

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4, SSSF, ANS....

In [ ]:
cnm_unsided.head()

#### Merging cnm_collapsing_unsided and cnm_unsided

In [ ]:
cnm_collapsed = pd.DataFrame()

In [ ]:
cnm_collapsed = pd.concat([cnm_unsided, cnm_collapsing_unsided], axis=1)

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4, SSSF, ANS....

In [ ]:
cnm_collapsed.head()

In [ ]:
print(retained_traits)

{'MIFR', 'PTBL', 'CIVR', 'OMBL', 'MFLoR'}


In [ ]:
# bring over columns from original cnm df whose opposite sides were removed due to low IO agreement

cnm_collapsed['MIF'] = cnm['MIFR']
cnm_collapsed['PTB'] = cnm['PTBL']
cnm_collapsed['CIV'] = cnm['CIVR']
cnm_collapsed['OMB'] = cnm['OMBL']
cnm_collapsed['MFLo'] = cnm['MFLoR']

In [ ]:
# bring over PHAR because it was accidentally missed because it ends in 'R'

cnm_collapsed['PHAR'] = cnm['PHAR']

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4, SSSF, ANS....

In [ ]:
cnm_collapsed.head()

In [ ]:
cnm.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 97 entries, 0 to 96
Data columns (total 85 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   SkelID       97 non-null     object 
 1   Collection   97 non-null     object 
 2   Sex          97 non-null     object 
 3   Age          97 non-null     float64
 4   Population   97 non-null     object 
 5   Population2  19 non-null     object 
 6   Population3  10 non-null     object 
 7   Population4  1 non-null      object 
 8   SONR         96 non-null     float64
 9   SONL         97 non-null     float64
 10  SOFR         96 non-null     float64
 11  SOFL         97 non-null     float64
 12  IFSR         94 non-null     float64
 13  IFSL         94 non-null     float64
 14  MIFR         96 non-null     float64
 15  ZFFR         97 non-null     float64
 16  ZFFL         97 non-null     float64
 17  CCOR         97 non-null     float64
 18  CCOL         96 non-null     float64
 19  SSSF      

In [ ]:
cnm_collapsed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 97 entries, 0 to 96
Data columns (total 58 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   SkelID       97 non-null     object 
 1   Collection   97 non-null     object 
 2   Sex          97 non-null     object 
 3   Age          97 non-null     float64
 4   Population   97 non-null     object 
 5   Population2  19 non-null     object 
 6   Population3  10 non-null     object 
 7   Population4  1 non-null      object 
 8   SSSF         96 non-null     float64
 9   ANS          95 non-null     float64
 10  INA          97 non-null     float64
 11  IOB          96 non-null     float64
 12  NAS          96 non-null     float64
 13  NAW          95 non-null     float64
 14  NBC          94 non-null     float64
 15  NBS          94 non-null     float64
 16  NFS          95 non-null     float64
 17  OBS          96 non-null     float64
 18  TPS          89 non-null     float64
 19  INCA      

##### Output data file: cnm_collapsed

In [ ]:
cnm_collapsed.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_collapsed.xlsx', index=False)

##### Old code

In [ ]:
'''
our_column_names = []                    # create an empty list for the column names
for t in list(cnm_VarsOnly.columns):
    if t in retained_traits:
      print(t)
      continue
    if t[-1] == 'L':                     # find all column names that end in L (double check that on Lefts end in L prior to this step)
        our_column_names.append(t[:-1])  # t[:-1] truncates the last letter (L or R) from the col name
print(our_column_names)
'''

MIFR
CIVR
PTBL
MFLoR
OMBL
['SON', 'SOF', 'IFS', 'ZFF', 'CCO', 'FOI', 'FSI', 'TYM', 'AUD', 'MT', 'NO', 'PZT', 'ZS', 'LBM', 'LBLa', 'PF', 'MFN', 'CRB', 'EPB', 'FTA', 'PNB', 'AST', 'HYP', 'APF', 'FF', 'MHB', 'MEN']


In [ ]:
'''
cnm_collapsed = pd.DataFrame()
for c in cnm_VarsOnly.columns:
    c_l = c[-1] == 'L'                                # loop through all Lefts in variable c_l
    c_r = c[-1] == 'R'                               # loop through all Rights in variable c_r
    #print(value)
    cnm_collapsed[c] = cnm_VarsOnly[c_l]
    if c_l in retained_traits or c_r in retained_traits:
        if c_r in retained_traits:
          cnm_collapsed[c] = cnm_VarsOnly[c_r]
          continue
        if c_l in retained_traits:
          continue                        # overwrite the base (unsided) column variables (c) with Left values (c_l)
    cnm_collapsed['mask'] = (cnm_VarsOnly[c].isnull())       # create a mask variable where all NaNs will be True
    cnm_collapsed.loc[cnm_collapsed['mask'], c] = cnm_VarsOnly[c_r] # for rows that have NaNs in 'mask', overwrite the corresponding value in c with the value from the Right side (c_r)
#print(cnm_VarsOnly[c].tail(15))                   # print the unsided columns tail to check for
#print(cnm_VarsOnly['SON'].head())
#print(cnm_VarsOnly['FOI'].tail(10))
'''

KeyError: ignored

In [ ]:
'''
cnm_collapsed = pd.DataFrame()
for c in cnm_VarsOnly:
    c_l = c +'L'                                # loop through all Lefts in variable c_l
    c_r = c + 'R'                               # loop through all Rights in variable c_r
    #print(value)
    cnm_collapsed[c] = cnm_VarsOnly[c_l]
    if c_l in retained_traits or c_r in retained_traits:
        if c_r in retained_traits:
          cnm_collapsed[c] = cnm_VarsOnly[c_r]
          continue
        if c_l in retained_traits:
          continue                        # overwrite the base (unsided) column variables (c) with Left values (c_l)
    cnm_collapsed['mask'] = (cnm_VarsOnly[c].isnull())       # create a mask variable where all NaNs will be True
    cnm_collapsed.loc[cnm_collapsed['mask'], c] = cnm_VarsOnly[c_r] # for rows that have NaNs in 'mask', overwrite the corresponding value in c with the value from the Right side (c_r)
#print(cnm_VarsOnly[c].tail(15))                   # print the unsided columns tail to check for
#print(cnm_VarsOnly['SON'].head())
#print(cnm_VarsOnly['FOI'].tail(10))
'''

KeyError: ignored

In [ ]:
'''
for c in our_column_names:
    c_l = c +'L'                                # loop through all Lefts in variable c_l
    c_r = c + 'R'                               # loop through all Rights in variable c_r
    #print(value)
    cnm_VarsOnly[c] = cnm_VarsOnly[c_l]
    if c_l in retained_traits or c_r in retained_traits:
        if c_r in retained_traits:
          cnm_collapsed[c] = cnm_VarsOnly[c_r]
          continue
        if c_l in retained_traits:
          continue                        # overwrite the base (unsided) column variables (c) with Left values (c_l)
    cnm_VarsOnly['mask'] = (cnm_VarsOnly[c].isnull())       # create a mask variable where all NaNs will be True
    cnm_VarsOnly.loc[cnm_VarsOnly['mask'], c] = cnm_VarsOnly[c_r] # for rows that have NaNs in 'mask', overwrite the corresponding value in c with the value from the Right side (c_r)
#print(cnm_VarsOnly[c].tail(15))                   # print the unsided columns tail to check for
#print(cnm_VarsOnly['SON'].head())
#print(cnm_VarsOnly['FOI'].tail(10))
'''

82    0
83    0
84    0
85    0
86    0
87    0
88    0
89    1
90    0
91    0
92    0
93    0
94    0
95    0
96    0
Name: MEN, dtype: int64


In [ ]:
'''
our_column_names = []                    # create an empty list for the column names
for c in list(cnm.columns):

    if c[-1] == 'L':                     # find all column names that end in L (double check that on Lefts end in L prior to this step)
        our_column_names.append(c)  # c[:-1] truncates the last letter (L or R) from the col name
print(our_column_names)
'''

['SONL', 'SOFL', 'IFSL', 'ZFFL', 'CCOL', 'FOIL', 'FSIL', 'PTBL', 'TYML', 'AUDL', 'MTL', 'NOL', 'PZTL', 'ZSL', 'LBML', 'LBLL', 'PFL', 'MFL', 'CRBL', 'EPBL', 'FTAL', 'PNBL', 'ASTL', 'OMBL', 'HYPL', 'APFL', 'FFL', 'MHBL', 'MENL']


In [ ]:
'''
cnm_collapsed = pd.DataFrame()
for c in cnm_VarsOnly.columns:
    c_l = c +'L'                                # loop through all Lefts in variable c_l
    c_r = c + 'R'                               # loop through all Rights in variable c_r
    #print(value)
    cnm_collapsed[c] = cnm_VarsOnly[c_l]
    if c_l in retained_traits or c_r in retained_traits:
      if c_r in retained_traits:
        cnm_collapsed[c] = cnm_VarsOnly[c_r]
        continue
      if c_l in retained_traits:
        continue                                                # overwrite the base (unsided) column variables (c) with Left values (c_l)
    cnm_collapsed['mask'] = (cnm_collapsed[c].isnull())       # create a mask variable where all NaNs will be True
    cnm_collapsed.loc[cnm_VarsOnly['mask'], c] = cnm_VarsOnly[c_r] # for rows that have NaNs in 'mask', overwrite the corresponding value in c with the value from the Right side (c_r)
    print(cnm_collapsed[c].tail(15))                   # print the unsided columns tail to check for
#print(cnm['SON'].head())
#print(cnm['FOI'].tail(10))
'''

KeyError: ignored

In [ ]:
'''
# collected the unsided names

unsided_names = []
for c in cnm.columns:
  if c[-1] == 'L' or c[-1] == 'R':
    continue
  unsided_names.append(c)
print(unsided_names)
'''

['SkelID', 'Collection', 'Sex', 'Age', 'Population', 'Population2', 'Population3', 'Population4', 'SSSF', 'ANS', 'INA', 'IOB', 'NAS', 'NAW', 'NBC', 'NBS', 'NFS', 'OBS', 'TPS', 'INCA', 'BREG', 'PALT', 'APIC', 'METO', 'MANT']


In [ ]:
'''
for c in unsided_names:
  cnm_collapsed[c] = cnm[c]
  '''

### Find numbers of males and females for each population

#### My data

In [ ]:
cnm2 = cnm_cleaned2.groupby('Collection').apply(lambda x: x['Sex'].value_counts())

The output for print(cnm2) was removed to protect PII. The output resembled the following:

* Sex          M     F
* Collection
* US           #     #
* Japan        #     #

In [ ]:
print(cnm2)

#### Comparative Dataset 1 (Comp1)

In [ ]:
Comp1_cnm = pd.read_excel('Comp1_cnm_for_analysis.xlsx')

The output for Comp1_cnm.head() was removed to protect PII. The output included the following columns:
* ID, Curator, Sex, Race, Birth Year, Age at Death, Lamb Oss Med R, Lamb Oss Med L, .....

In [ ]:
Comp1_cnm.head()

In [ ]:
Comp1_cnm2 = Comp1_cnm.groupby('Race').apply(lambda x: x['Sex'].value_counts())
Comp1_cnm2

Race   
B     M    36
      F    24
EA    M     3
      F     1
H     M    26
      F     3
W     M    86
      F    57
W/H   M     1
W/NA  M     1
Name: Sex, dtype: int64

#### Comparative Dataset 2 (Comp2)

In [ ]:
Comp2_cnm = pd.read_excel('Comp2_cnm_for_analysis.xlsx')

The output from Comp2_cnm.head() has been removed to protect PII. The output included the following columns:
* file.name, GP1, GP2, GP3, GP4, GP4, Site, CatalogNo, Museum, AgeC, AgeY, Sex, DeformOriginal, DeformGrade, DeformLRS, METO, APIC, INCA, OMBL, OMBR, ASTL, ASTR, ....

In [ ]:
Comp2_cnm.head()

In [ ]:
Comp2_cnm2 = Comp2_cnm.groupby('GP3').apply(lambda x: x['Sex'].value_counts())
print(Comp2_cnm2)

GP3     
AL   4.0    216
     0.0    212
AT   4.0     98
     0.0     87
AU   0.0     30
           ... 
TF   4.0      5
USA  0.0     34
     4.0     28
W    0.0    163
     4.0    114
Name: Sex, Length: 84, dtype: int64


In [ ]:
Comp2_cnm2

GP3      
AL    4.0    216
      0.0    212
AT    4.0     98
      0.0     87
AU    0.0     30
      4.0     23
C     0.0    234
      4.0    145
CAM   0.0      2
CAN   0.0    129
      4.0    107
CAR   4.0    190
      0.0    182
CH    0.0     25
      4.0      8
CHAT  0.0     11
      4.0      9
CHN   0.0     55
      4.0     14
EAR   4.0    212
      0.0    162
GAB   0.0      3
      4.0      3
GHA   0.0     18
      4.0     14
HK    0.0    109
      4.0     92
ILL   0.0     51
      4.0     42
IND   0.0     72
      4.0     55
KEN   4.0     17
      0.0      8
LIB   4.0      2
MON   0.0     35
      4.0     25
MQ    0.0     38
      4.0     20
N     0.0    118
      4.0     96
NAL   0.0    223
      4.0    212
NFL   0.0     21
      4.0     18
NIG   0.0     15
      4.0     14
NMV   0.0    236
      4.0    168
NN    0.0     31
      4.0     21
NPC   0.0    266
      4.0    198
NZ    0.0     24
      4.0     22
ONT   4.0     35
      0.0     32
PEC   4.0     84
      0.0     76
PLN 

In [ ]:
Comp2_cnm3 = Comp2_cnm.groupby('GP2').apply(lambda x: x['Sex'].value_counts())
print(Comp2_cnm3)

GP2     
AM   0.0      34
     4.0      28
AR   4.0    1117
     0.0    1039
ARM  0.0      70
     4.0      59
BR   0.0     129
     4.0     107
BV   0.0       4
     4.0       3
CZ   0.0      10
     4.0       3
E    4.0      48
     0.0      24
EU   0.0       4
FR   0.0       1
GE   0.0       5
     4.0       2
HU   0.0      42
     4.0      21
IC   0.0      22
     4.0      22
IN   0.0      72
     4.0      55
IT   0.0      44
     4.0      44
JA   0.0     648
     4.0     463
N    0.0      48
     4.0      32
NE   0.0     104
     4.0      95
NEA  0.0     186
     4.0     131
NW   0.0    1064
     4.0     886
RU   0.0      13
S    0.0      44
     4.0      21
SA   0.0      44
     4.0      15
SW   4.0      84
     0.0      76
W    0.0      45
     4.0      38
Name: Sex, dtype: int64


In [ ]:
from google.colab import  drive
drive.mount('/drive')

Mounted at /drive


In [ ]:
Comp2_cnm2.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/Comp2_sex_byGP3.xlsx')

In [ ]:
Comp2_cnm3.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/Comp2_sex_byGP2.xlsx')

#### Comparative Dataset 3 (Comp3)

In [ ]:
Comp3_cnm = pd.read_excel('Comp3_for_analysis.xlsx')

The output for Comp2_cnm.head() was removed to protect PII. The output included the following columns:
* ID, Provenience, Ancestry, Sex, Age, ANS...

In [ ]:
Comp3_cnm.head()

In [ ]:
Comp3_cnm2 = Comp3_cnm.groupby('Ancestry').apply(lambda x: x['Sex'].value_counts())
print(Comp3_cnm2)

Ancestry             
AmericanBlack  Male      56
               Female    44
AmericanWhite  Male      55
               Female    44
Amerindian     Male       1
Hispanic       Male      73
               Female    26
Name: Sex, dtype: int64


### EDA with d-tale

In [ ]:
import dtale.app as dtale_app

dtale_app.USE_COLAB = True

In [ ]:
dtale.show(cnm_collapsed)

In [ ]:
cnm_collapsed.describe()

Age       SSSF        ANS        INA        IOB        NAS        NAW        NBC        NBS        NFS        OBS        TPS       INCA       BREG       PALT       APIC       METO       MANT        SON        SOF        IFS        ZFF        CCO        FOI        FSI        PTB        TYM        AUD         MT         NO        PZT         ZS        LBM       LBLa         PF         MF        CRB        EPB        FTA        PNB        AST        OMB        HYP        APF         FF        MHB        MEN        MIF        CIV       MFLo       PHAR
count   97.000000  96.000000  95.000000  97.000000  96.000000  96.000000  95.000000  94.000000  94.000000  95.000000  96.000000  89.000000  80.000000  83.000000  95.000000  41.000000  96.000000  91.000000  97.000000  97.000000  95.000000  97.000000  97.000000  96.000000  96.000000  97.000000  97.000000  97.000000  97.000000  94.000000  97.000000  92.000000  64.000000  80.000000  95.000000  96.000000  81.000000  70.000000  81.000000  94.000000  88.000000  64.000000  96.000000  97.000000  97.000000  97.000000  96.000000  96.000000  95.000000  94.000000  95.000000
mean    61.371134   1.229167   2.305263   3.226804   1.906250   1.447917   1.968421   1.457447   2.457447   2.200000   1.645833   2.067416   0.012500   0.024096   0.800000   0.243902   0.020833   0.384615   0.855670   0.597938   1.031579   1.628866   0.597938   0.041667   0.145833   0.206186   0.175258   0.030928   1.773196   0.542553   1.896907   1.641304   0.250000   0.375000   0.726316   1.614583   0.358025   0.142857   0.012346   0.308511   0.102273   0.171875   0.593750   1.206186   0.185567   0.340206   0.062500   0.604167   0.915789   2.425532   0.252632
std     23.882228   0.606616   0.715686   1.113414   0.697033   0.559507   0.398204   1.404026   0.698153   1.154086   0.820515   1.009025   0.111803   0.154281   0.402122   0.434769   0.204124   0.489200   0.645331   0.687080   0.818049   0.564941   0.492861   0.200875   0.383314   0.455022   0.478938   0.226090   0.756984   0.500857   0.809894   0.585193   0.436436   0.487177   0.514515   0.933243   0.482407   0.352454   0.111111   0.464355   0.304743   0.380254   1.192769   0.923514   0.485839   0.518142   0.243332   0.888276   0.767188   1.433016   0.436827
min     17.000000   0.000000   0.000000   1.000000   1.000000   1.000000   1.000000   0.000000   1.000000   0.000000   1.000000   1.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000
25%     40.000000   1.000000   2.000000   2.000000   1.000000   1.000000   2.000000   0.000000   2.000000   1.000000   1.000000   1.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000   1.000000   1.000000   0.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000
50%     68.000000   1.000000   2.000000   3.000000   2.000000   1.000000   2.000000   1.000000   2.000000   2.000000   1.000000   2.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   1.000000   0.000000   1.000000   2.000000   1.000000   0.000000   0.000000   0.000000   0.000000   0.000000   2.000000   1.000000   2.000000   2.000000   0.000000   0.000000   1.000000   2.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   1.000000   0.000000   0.000000   0.000000   0.000000   1.000000   2.000000   0.000000
75%     82.000000   1.000000   3.00000

In [ ]:
cnm_collapsed.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 97 entries, 0 to 96
Data columns (total 58 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   SkelID       97 non-null     object 
 1   Collection   97 non-null     object 
 2   Sex          97 non-null     object 
 3   Age          97 non-null     float64
 4   Population   97 non-null     object 
 5   Population2  19 non-null     object 
 6   Population3  10 non-null     object 
 7   Population4  1 non-null      object 
 8   SSSF         96 non-null     float64
 9   ANS          95 non-null     float64
 10  INA          97 non-null     float64
 11  IOB          96 non-null     float64
 12  NAS          96 non-null     float64
 13  NAW          95 non-null     float64
 14  NBC          94 non-null     float64
 15  NBS          94 non-null     float64
 16  NFS          95 non-null     float64
 17  OBS          96 non-null     float64
 18  TPS          89 non-null     float64
 19  INCA      

## EDA with Sweetviz

In [ ]:
pip install sweetviz

     |████████████████████████████████| 15.1 MB 4.2 MB/s 


In [ ]:
import sweetviz as sv

In [ ]:
#type(cnm['SONR'])

pandas.core.series.Series

In [ ]:
#type(cnm)

pandas.core.frame.DataFrame

In [ ]:
#cnm['SONR'] = pd.to_numeric(cnm['SONR'], errors='coerce')

In [ ]:
advert_report = sv.analyze(cnm_collapsed)

                                             |          | [  0%]   00:00 -> (? left)

In [ ]:
advert_report.show_html('cnm_collapsed.html', open_browser=False)

Report cnm_collapsed.html was generated.


In [ ]:
import IPython

The IPython.display.HTML('cnm_collapsed.html) was removed to protect PII.

In [ ]:
IPython.display.HTML('cnm_collapsed.html')

## Drop columns with > 50% missing data

In [ ]:
cnm_collapsed = pd.read_excel('cnm_collapsed.xlsx')

In [ ]:
cnm_cleaned2 = pd.DataFrame(cnm_collapsed)

The output for the following code cell was removed to protect PII.

In [ ]:
cnm_cleaned2.head()

In [ ]:
column_names = cnm_cleaned2.columns                                        # create a list of the column names in cnm_cleaned2 df

for i in range(8, len(cnm_cleaned2.columns)):                              # starting at column 8 and going until the end (len) of the df
  #print(cnm_cleaned2.columns[i])
  percent_missing = (cnm_cleaned2[column_names[i]].isnull().sum() * 100 / len(cnm_cleaned2))    # calculate the percent missing data for each column in the df
  if percent_missing > 50:                                                       # if the percentage of missing data is > greater than 50
    print(column_names[i], '\t', percent_missing)                                 # print the column name and the percent missing data for those columns
    del cnm_cleaned2[column_names[i]]                                      # delete the columns with >= 50% missing data
  #print(column_names[i], '\t', (cnm_cleaned2[column_names[i]].isnull().sum() * 100 / len(cnm_cleaned2)))

APIC 	 57.7319587628866


Removed APIC because it had > 50% missing data

In [ ]:
cnm_cleaned2.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_cleaned2.xlsx', index=False)

## Revisualize with Sweetviz

In [ ]:
cnm_cleaned2 = pd.read_excel('cnm_cleaned2.xlsx')

The output for the following code cell was removed to protect PII.

In [ ]:
cnm_cleaned2.head()

In [ ]:
advert_report = sv.analyze(cnm_cleaned2)

                                             |          | [  0%]   00:00 -> (? left)

In [ ]:
advert_report.show_html('cnm_cleaned2.html', open_browser=False)

Report cnm_cleaned2.html was generated.


In [ ]:
import IPython

The output for the following code cell was removed to protect PII.

In [ ]:
IPython.display.HTML('cnm_cleaned2.html')

# **Sexual dimorphism**
chi-square for SD and to calculate trait frequencies

In [ ]:
from scipy.stats import chisquare

In [ ]:
cnm_cleaned2 = pd.read_excel('cnm_cleaned2.xlsx')

The output for the following code cell was removed to protect PII.

In [ ]:
cnm_cleaned2.head()

In [ ]:
cnm_cleaned2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 0 to 96
Data columns (total 57 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   SkelID       97 non-null     int64  
 1   Collection   97 non-null     object 
 2   Sex          97 non-null     object 
 3   Age          97 non-null     int64  
 4   Population   97 non-null     object 
 5   Population2  19 non-null     object 
 6   Population3  10 non-null     object 
 7   Population4  1 non-null      object 
 8   SSSF         96 non-null     float64
 9   ANS          95 non-null     float64
 10  INA          97 non-null     int64  
 11  IOB          96 non-null     float64
 12  NAS          96 non-null     float64
 13  NAW          95 non-null     float64
 14  NBC          94 non-null     float64
 15  NBS          94 non-null     float64
 16  NFS          95 non-null     float64
 17  OBS          96 non-null     float64
 18  TPS          89 non-null     float64
 19  INCA      

### Chi-square using Researchpy package
https://researchpy.readthedocs.io/

In [ ]:
!pip install researchpy

In [ ]:
import researchpy

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [ ]:
cnm_VarsOnly = pd.DataFrame(cnm_cleaned2)

The output for the following code cell was removed to protect PII.

In [ ]:
cnm_VarsOnly.head()

In [ ]:
del cnm_VarsOnly['SkelID']
del cnm_VarsOnly['Collection']
del cnm_VarsOnly['Age']
del cnm_VarsOnly['Population']
del cnm_VarsOnly['Population2']
del cnm_VarsOnly['Population3']
del cnm_VarsOnly['Population4']

The output for the following code cell was removed to protect PII.

In [ ]:
cnm_VarsOnly.head()

##### chi-square individual results for SSSF



In [ ]:
# Output includes the crosstable, results, and expected table
crosstab, res, exp = researchpy.crosstab(cnm_VarsOnly['Sex'], cnm_VarsOnly['SSSF'], test='chi-square', expected_freqs=True)

In [ ]:
res     # results shown as pandas df

Chi-square test  results
0  Pearson Chi-square ( 3.0) =    1.8715
1                    p-value =    0.5995
2                 Cramer's V =    0.1396

In [ ]:
exp

SSSF                               
SSSF       0.0        1.0       2.0       3.0
Sex                                          
F     0.447917  35.833333  3.135417  3.583333
M     0.552083  44.166667  3.864583  4.416667

In [ ]:
crosstab

SSSF                
SSSF  0.0 1.0 2.0 3.0 All
Sex                      
F       1  35   4   3  43
M       0  45   3   5  53
All     1  80   7   8  96

##### Manual chi-square calculations

In [ ]:
myCrosstable = pd.crosstab(cnm_VarsOnly['Sex'], cnm_VarsOnly['SSSF'])
myCrosstable

SSSF  0.0  1.0  2.0  3.0
Sex                     
F       1   35    4    3
M       0   45    3    5

In [ ]:
colTotals = myCrosstable.sum()
rowTotals = myCrosstable.sum(axis=1)

colTotals, rowTotals

(SSSF
 0.0     1
 1.0    80
 2.0     7
 3.0     8
 dtype: int64, Sex
 F    43
 M    53
 dtype: int64)

In [ ]:
nCols = len(colTotals)
nRows = len(rowTotals)

nCols, nRows

(4, 2)

In [ ]:
N = rowTotals.sum()
N

96

In [ ]:
for i in range(nRows):
    for j in range(nCols):
        E = rowTotals[i] * colTotals[j] / N
        print(E)

0.4479166666666667
35.833333333333336
3.1354166666666665
3.5833333333333335
0.5520833333333334
44.166666666666664
3.8645833333333335
4.416666666666667


In [ ]:
chiVal = 0
for i in range(nRows):
    for j in range(nCols):
        E = rowTotals[i] * colTotals[j] / N
        O = myCrosstable.iloc[i,j]
        chiVal = chiVal + (O - E)**2 / E
chiVal

1.8714975239766818

In [ ]:
df = (nRows - 1) * (nCols - 1)
df

3

In [ ]:
# !pip install scipy
from scipy.stats import chi2

In [ ]:
chi2.cdf(chiVal, df)

0.4004987970229535

In [ ]:
1 - chi2.cdf(chiVal, df)

0.5995012029770466

In [ ]:
chi2.sf(chiVal, df)

0.5995012029770466

In [ ]:
crosstab, res, exp = researchpy.crosstab(cnm_cleaned2['Sex'], cnm_cleaned2['SSSF'], test='chi-square', expected_freqs=True)
print(crosstab)
print(res)
print(exp)

     SSSF                
SSSF  0.0 1.0 2.0 3.0 All
Sex                      
F       1  35   4   3  43
M       0  45   3   5  53
All     1  80   7   8  96
                Chi-square test  results
0  Pearson Chi-square ( 3.0) =    1.8715
1                    p-value =    0.5995
2                 Cramer's V =    0.1396
          SSSF                               
SSSF       0.0        1.0       2.0       3.0
Sex                                          
F     0.447917  35.833333  3.135417  3.583333
M     0.552083  44.166667  3.864583  4.416667


##### chi-square for all results

In [ ]:
#column_names = cnm_VarsOnly.columns

for c in cnm_VarsOnly.columns:
    crosstab, res, exp = researchpy.crosstab(cnm_VarsOnly['Sex'], cnm_VarsOnly[c], test='chi-square', expected_freqs=True)
    print(crosstab)
    print(res)
    print(exp)

    Sex        
Sex   F   M All
Sex            
F    43   0  43
M     0  54  54
All  43  54  97
                Chi-square test  results
0  Pearson Chi-square ( 1.0) =      97.0
1                    p-value =       0.0
2               Cramer's phi =       1.0
           Sex           
Sex          F          M
Sex                      
F    19.061856  23.938144
M    23.938144  30.061856
     SSSF                
SSSF  0.0 1.0 2.0 3.0 All
Sex                      
F       1  35   4   3  43
M       0  45   3   5  53
All     1  80   7   8  96
                Chi-square test  results
0  Pearson Chi-square ( 3.0) =    1.8715
1                    p-value =    0.5995
2                 Cramer's V =    0.1396
          SSSF                               
SSSF       0.0        1.0       2.0       3.0
Sex                                          
F     0.447917  35.833333  3.135417  3.583333
M     0.552083  44.166667  3.864583  4.416667
    ANS                
ANS 0.0 1.0 2.0 3.0 All
Sex         

##### chi-square for loop

In [ ]:
# to prep the chi-square for loop, find out: what type of data structure is res?

print(type(res))

# since res is a dataframe, check the parts of the res df

<class 'pandas.core.frame.DataFrame'>


In [ ]:
res.columns

Index(['Chi-square test', 'results'], dtype='object')

In [ ]:
print(res)

                Chi-square test  results
0  Pearson Chi-square ( 1.0) =    1.5404
1                    p-value =    0.2146
2               Cramer's phi =    0.1273


In [ ]:
res['Chi-square test']

0    Pearson Chi-square ( 1.0) = 
1                      p-value = 
2                 Cramer's phi = 
Name: Chi-square test, dtype: object

In [ ]:
res['results']

0    1.5404
1    0.2146
2    0.1273
Name: results, dtype: float64

In [ ]:
res['results'][0]

1.5404

In [ ]:
print(res['results'][0], '\t', res['results'][1], '\t', res['results'][2])

1.5404 	 0.2146 	 0.1273


In [ ]:
#column_names = cnm_VarsOnly.columns
column_names = ('Trait Code', 'Pearson Chi-square', 'p-value', 'Cramer\'s phi')
chi_sq_df = pd.DataFrame(columns = column_names)
for c in cnm_VarsOnly.columns:
    crosstab, res, exp = researchpy.crosstab(cnm_VarsOnly['Sex'], cnm_VarsOnly[c], test='chi-square', expected_freqs=True)
    #print(crosstab)
    #print(type(res))
    #print(exp)
    print(c, res['results'][0], '\t', res['results'][1], '\t', res['results'][2])
    chi_sq_df.loc[c] = [c, res['results'][0], res['results'][1], res['results'][2]]

Sex 97.0 	 0.0 	 1.0
SSSF 1.8715 	 0.5995 	 0.1396
ANS 5.9 	 0.1166 	 0.2492
INA 2.1673 	 0.705 	 0.1495
IOB 0.8755 	 0.6455 	 0.0955
NAS 1.4545 	 0.4832 	 0.1231
NAW 0.7643 	 0.6824 	 0.0897
NBC 8.7693 	 0.0325 	 0.3054
NBS 6.3252 	 0.0968 	 0.2594
NFS 5.7946 	 0.215 	 0.247
OBS 4.6145 	 0.0995 	 0.2192
TPS 3.8044 	 0.2834 	 0.2068
INCA 0.8285 	 0.3627 	 0.1018
BREG 1.422 	 0.2331 	 0.1309
PALT 2.7462 	 0.0975 	 0.17
METO 0.8199 	 0.3652 	 0.0924
MANT 0.0397 	 0.8421 	 0.0209
SON 2.242 	 0.5237 	 0.152
SOF 1.7459 	 0.4177 	 0.1342
IFS 6.9224 	 0.0314 	 0.2699
ZFF 2.4725 	 0.4803 	 0.1597
CCO 0.0879 	 0.7668 	 0.0301
FOI 0.6612 	 0.4162 	 0.083
FSI 1.9027 	 0.3862 	 0.1408
PTB 1.644 	 0.4396 	 0.1302
TYM 1.9684 	 0.3737 	 0.1425
AUD 1.6261 	 0.4435 	 0.1295
MT 1.6095 	 0.6572 	 0.1288
NO 0.5539 	 0.4567 	 0.0768
PZT 2.8608 	 0.4136 	 0.1717
ZS 1.5211 	 0.4674 	 0.1286
LBM 0.5339 	 0.465 	 0.0913
LBLa 0.6684 	 0.4136 	 0.0914
PF 4.4818 	 0.1064 	 0.2172
MF 12.3307 	 0.0063 	 0.3584
CRB 

In [ ]:
chi_sq_df.head()

Trait Code  Pearson Chi-square  p-value  Cramer's phi
Sex         Sex             97.0000   0.0000        1.0000
SSSF       SSSF              1.8715   0.5995        0.1396
ANS         ANS              5.9000   0.1166        0.2492
INA         INA              2.1673   0.7050        0.1495
IOB         IOB              0.8755   0.6455        0.0955

In [ ]:
codes_names = pd.read_excel('trait codes and names_unsided.xlsx')
codes_names.head()

Trait Code                      Trait Name
0        SON              Supraorbital notch
1        SOF            Supraorbital foramen
2        IFS             Infraorbital suture
3        MIF  Multiple infraorbital foramina
4        ZFF      Zygomatico-facial foramina

In [ ]:
chi_square_results = pd.merge(chi_sq_df, codes_names, on='Trait Code')
chi_square_results.head()

Trait Code  Pearson Chi-square  p-value  Cramer's phi                           Trait Name
0       SSSF              1.8715   0.5995        0.1396  Flexure of superior sagittal sulcus
1        ANS              5.9000   0.1166        0.2492                 Anterior nasal spine
2        INA              2.1673   0.7050        0.1495              Inferior nasal aperture
3        IOB              0.8755   0.6455        0.0955                 Interorbital breadth
4        NAS              1.4545   0.4832        0.1231                 Nasal aperture shape

#### Output data file:

In [ ]:
chi_square_results.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_SD_chi_square.xlsx', index=False)

# **Intertrait correlations**
Kendall's tau b

In [ ]:
#cnm_tauB = pd.DataFrame(cnm_VarsOnly)

In [ ]:
import scipy.stats as stats

The output for the following code cell was removed to protect PII. The output contained the following columns:
* Sex, SSSF, ANS.....

In [ ]:
cnm_VarsOnly.head()

In [ ]:
cnm_tauB = pd.DataFrame(cnm_VarsOnly)

The output for the following code cell was removed to protect PII. The output contained the following columns:
* Sex, SSSF, ANS.....

In [ ]:
cnm_tauB.head()

In [ ]:
cnm_tauB = cnm_tauB.dropna()

In [ ]:
cnm_tauB.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 39 entries, 0 to 96
Data columns (total 50 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Sex     39 non-null     object 
 1   SSSF    39 non-null     float64
 2   ANS     39 non-null     float64
 3   INA     39 non-null     int64  
 4   IOB     39 non-null     float64
 5   NAS     39 non-null     float64
 6   NAW     39 non-null     float64
 7   NBC     39 non-null     float64
 8   NBS     39 non-null     float64
 9   NFS     39 non-null     float64
 10  OBS     39 non-null     float64
 11  TPS     39 non-null     float64
 12  INCA    39 non-null     float64
 13  BREG    39 non-null     float64
 14  PALT    39 non-null     float64
 15  METO    39 non-null     float64
 16  MANT    39 non-null     float64
 17  SON     39 non-null     int64  
 18  SOF     39 non-null     int64  
 19  IFS     39 non-null     float64
 20  ZFF     39 non-null     int64  
 21  CCO     39 non-null     int64  
 22  FOI 

In [ ]:
tau, p_value = stats.kendalltau(cnm_tauB['Sex'], cnm_tauB['SSSF'])
print(tau)
print(p_value)

-0.02749806661015982
0.8639459980780579


In [ ]:
tau, p_value = stats.kendalltau(cnm_tauB['Sex'], cnm_tauB['ANS'])
print(tau)
print(p_value)

0.01703208390548536
0.9124937972451566


In [ ]:
tau, p_value = stats.kendalltau(cnm_tauB['Sex'], cnm_tauB['INCA'])
print(tau)
print(p_value)

nan
nan


In [ ]:

'''# Kendall's tau b with one column fixed and the other columns iterating
# NA values were dropped from the cnm_tauB df (only 11 individuals in this analysis)

#len(cnm_tauB[cnm_tauB.columns[3]])
column_names = ('Trait Code', 'tau B statistic', 'p-value')
my_tauB_df2 = pd.DataFrame(columns = column_names)
for c in cnm_tauB.columns:
    tau, p_value = stats.kendalltau(cnm_tauB['Sex'], cnm_tauB[c])
    print(c,'\t', tau, '\t', p_value)
    #df3 = {'Trait Code': [c], 'tau B statistic': tau[0], 'p-value': p_value[1]}
    my_tauB_df2.loc[c] = [c, tau, p_value]
    '''

Sex 	 1.0 	 7.074463098970675e-10
SSSF 	 -0.02749806661015982 	 0.8639459980780579
ANS 	 0.01703208390548536 	 0.9124937972451566
INA 	 0.0915507056838374 	 0.538684567676416
IOB 	 -0.13562205717103817 	 0.3793098072587159
NAS 	 -0.1725123787127004 	 0.28197106580647047
NAW 	 0.26086956521739135 	 0.10131105093820252
NBC 	 0.08318015020356473 	 0.5830650177135555
NBS 	 0.3259710630759294 	 0.0367841744519069
NFS 	 -0.3935623258746016 	 0.009375343000875485
OBS 	 -0.14264222271862334 	 0.3626425790575063
TPS 	 -0.09619325738134539 	 0.5200146809141549
INCA 	 nan 	 nan
BREG 	 nan 	 nan
PALT 	 -0.20755395563871767 	 0.20073916039155448
METO 	 0.13530201829834768 	 0.4042484947394712
MANT 	 0.25425818448750476 	 0.117032270472401
SON 	 0.203804347826087 	 0.19638082349986263
SOF 	 -0.12289192304282605 	 0.42743208750953476
IFS 	 -0.2545887739877383 	 0.09886741948203089
ZFF 	 -0.0493849928029914 	 0.755621721223902
CCO 	 0.19781414201873612 	 0.22268927373703729
FOI 	 -0.04241878948682547 

In [ ]:
#my_tauB_df2.head()

Trait Code  tau B statistic       p-value
Sex         Sex         1.000000  7.074463e-10
SSSF       SSSF        -0.027498  8.639460e-01
ANS         ANS         0.017032  9.124938e-01
INA         INA         0.091551  5.386846e-01
IOB         IOB        -0.135622  3.793098e-01

In [ ]:
codes_names = pd.read_excel('trait codes and names_unsided.xlsx')
codes_names.head()

Trait Code                      Trait Name
0        SON              Supraorbital notch
1        SOF            Supraorbital foramen
2        IFS             Infraorbital suture
3        MIF  Multiple infraorbital foramina
4        ZFF      Zygomatico-facial foramina

In [ ]:
tauB_results_named = pd.merge(my_tauB_df2, codes_names, on='Trait Code')
tauB_results_named.head()

Trait Code  tau B statistic   p-value                           Trait Name
0       SSSF        -0.027498  0.863946  Flexure of superior sagittal sulcus
1        ANS         0.017032  0.912494                 Anterior nasal spine
2        INA         0.091551  0.538685              Inferior nasal aperture
3        IOB        -0.135622  0.379310                 Interorbital breadth
4        NAS        -0.172512  0.281971                 Nasal aperture shape

In [ ]:
tauB_results_named = tauB_results_named.sort_values(by='tau B statistic', ascending=True)

In [ ]:
tauB_results_named.head()

Trait Code  tau B statistic   p-value                           Trait Name
0       SSSF        -0.027498  0.863946  Flexure of superior sagittal sulcus
1        ANS         0.017032  0.912494                 Anterior nasal spine
2        INA         0.091551  0.538685              Inferior nasal aperture
3        IOB        -0.135622  0.379310                 Interorbital breadth
4        NAS        -0.172512  0.281971                 Nasal aperture shape

#### tau B for loop

In [ ]:
trait_names = ['Trait Code']
trait_names.extend(cnm_tauB.columns)
print(trait_names)
my_tauB_df = pd.DataFrame(columns = trait_names)

['Trait Code', 'Sex', 'SSSF', 'ANS', 'INA', 'IOB', 'NAS', 'NAW', 'NBC', 'NBS', 'NFS', 'OBS', 'TPS', 'INCA', 'BREG', 'PALT', 'METO', 'MANT', 'SON', 'SOF', 'IFS', 'ZFF', 'CCO', 'FOI', 'FSI', 'PTB', 'TYM', 'AUD', 'MT', 'NO', 'PZT', 'ZS', 'LBM', 'LBLa', 'PF', 'MF', 'CRB', 'EPB', 'FTA', 'PNB', 'AST', 'OMB', 'HYP', 'APF', 'FF', 'MHB', 'MEN', 'MIF', 'CIV', 'MFLo', 'PHAR']


In [ ]:
my_tauB_df.head()

Empty DataFrame
Columns: [Trait Code, Sex, SSSF, ANS, INA, IOB, NAS, NAW, NBC, NBS, NFS, OBS, TPS, INCA, BREG, PALT, METO, MANT, SON, SOF, IFS, ZFF, CCO, FOI, FSI, PTB, TYM, AUD, MT, NO, PZT, ZS, LBM, LBLa, PF, MF, CRB, EPB, FTA, PNB, AST, OMB, HYP, APF, FF, MHB, MEN, MIF, CIV, MFLo, PHAR]
Index: []

In [ ]:
# Kendall's tau b
# NA values were dropped from the cnm_tauB df (only 11 individuals in this analysis)

trait_names = ['Trait Code']
trait_names.extend(cnm_tauB.columns)
my_tauB_df = pd.DataFrame(columns = trait_names)

for i in range(len(cnm_tauB.columns)):
    temp_df = pd.DataFrame()
    temp_list = []
    temp_list.append(cnm_tauB.columns[i])

    for j in range(len(cnm_tauB.columns)):
        tau, p_value = stats.kendalltau(cnm_tauB[cnm_tauB.columns[i]], cnm_tauB[cnm_tauB.columns[j]])
        print(c,'\t', tau, '\t', p_value)
        temp_list.append(tau)

    print(len(my_tauB_df.columns), '\t', len(temp_list))
    my_tauB_df.loc[i] = temp_list#[c, tau, p_value]

PHAR 	 1.0 	 7.074463098970675e-10
PHAR 	 -0.02749806661015982 	 0.8639459980780579
PHAR 	 0.01703208390548536 	 0.9124937972451566
PHAR 	 0.0915507056838374 	 0.538684567676416
PHAR 	 -0.13562205717103817 	 0.3793098072587159
PHAR 	 -0.1725123787127004 	 0.28197106580647047
PHAR 	 0.26086956521739135 	 0.10131105093820252
PHAR 	 0.08318015020356473 	 0.5830650177135555
PHAR 	 0.3259710630759294 	 0.0367841744519069
PHAR 	 -0.3935623258746016 	 0.009375343000875485
PHAR 	 -0.14264222271862334 	 0.3626425790575063
PHAR 	 -0.09619325738134539 	 0.5200146809141549
PHAR 	 nan 	 nan
PHAR 	 nan 	 nan
PHAR 	 -0.20755395563871767 	 0.20073916039155448
PHAR 	 0.13530201829834768 	 0.4042484947394712
PHAR 	 0.25425818448750476 	 0.117032270472401
PHAR 	 0.203804347826087 	 0.19638082349986263
PHAR 	 -0.12289192304282605 	 0.42743208750953476
PHAR 	 -0.2545887739877383 	 0.09886741948203089
PHAR 	 -0.0493849928029914 	 0.755621721223902
PHAR 	 0.19781414201873612 	 0.22268927373703729
PHAR 	 -0.0

In [ ]:
my_tauB_df.head()

Trait Code       Sex      SSSF       ANS       INA       IOB       NAS       NAW       NBC       NBS       NFS       OBS       TPS  INCA  BREG      PALT      METO      MANT       SON       SOF       IFS       ZFF       CCO       FOI       FSI       PTB       TYM  AUD        MT        NO       PZT        ZS       LBM      LBLa        PF        MF       CRB       EPB       FTA       PNB       AST       OMB       HYP       APF        FF       MHB       MEN       MIF       CIV      MFLo      PHAR
0        Sex  1.000000 -0.027498  0.017032  0.091551 -0.135622 -0.172512  0.260870  0.083180  0.325971 -0.393562 -0.142642 -0.096193   NaN   NaN -0.207554  0.135302  0.254258  0.203804 -0.122892 -0.254589 -0.049385  0.197814 -0.042419  0.211163  0.138675  0.019587  NaN -0.069001  0.249352  0.099293  0.102531 -0.092685  0.165508 -0.105125  0.347087 -0.061679 -0.061679  0.135302  0.012244 -0.346109  0.240772 -0.005398  0.118071 -0.146533  0.169750  0.007996  0.281468 -0.247670  0.068016  0.147442
1       SSSF -0.027498  1.000000  0.052321 -0.093208  0.036116 -0.225032  0.000000 -0.002923 -0.199720 -0.119361  0.009842 -0.076401   NaN   NaN -0.115926 -0.074876 -0.039651 -0.089369  0.173735  0.009029  0.159639 -0.052129 -0.107312  0.365509 -0.118212  0.090842  NaN  0.015048 -0.044089  0.039662  0.162116  0.406146 -0.066993  0.075011 -0.215322 -0.156038 -0.156038 -0.074876  0.038720 -0.133243  0.120553 -0.133152 -0.357266  0.448293  0.090842 -0.177003 -0.117491 -0.100250 -0.068234  0.104010
2        ANS  0.017032  0.052321  1.000000  0.325899 -0.213046  0.176190  0.087594 -0.066204  0.022803  0.063885  0.188100 -0.454694   NaN   NaN  0.248597 -0.075718  0.311867  0.330909 -0.381895 -0.208785 -0.147398 -0.103322 -0.108520 -0.126051 -0.121470 -0.008769  NaN -0.266307  0.043214 -0.037435  0.102007  0.097810  0.165980  0.002414 -0.174196  0.378705  0.110456 -0.249871  0.301500  0.017966  0.170673  0.036252  0.035240 -0.262412  0.087689 -0.157515 -0.080649  0.086488 -0.069302  0.058393
3        INA  0.091551 -0.093208  0.325899  1.000000 -0.015641 -0.110748  0.163749  0.024682  0.054409  0.085105  0.179015 -0.117647   NaN   NaN -0.019935  0.250157 -0.088724  0.221061 -0.184146 -0.250260  0.038326 -0.108366 -0.009959 -0.018265 -0.029727  0.024142  NaN -0.246349  0.099144  0.176065  0.128719 -0.089761  0.078881 -0.163907  0.143513  0.209974  0.166531 -0.180669  0.256568  0.193726  0.090680  0.228447  0.066583 -0.165214  0.061696  0.065706 -0.050884  0.028492  0.175380  0.433368
4        IOB -0.135622  0.036116 -0.213046 -0.015641  1.000000  0.139250  0.348969 -0.299421 -0.336705  0.078593 -0.068126  0.303607   NaN   NaN -0.545200  0.170300 -0.375112  0.085656  0.107604  0.137500  0.245034 -0.026462  0.074283  0.165431 -0.042234  0.237240  NaN  0.185417  0.058691  0.210487  0.224440 -0.063764 -0.130427  0.099127 -0.348429 -0.192879 -0.104155 -0.066639 -0.206382  0.021960 -0.320620  0.233970 -0.334472  0.280478 -0.002858 -0.206541 -0.064078 -0.045540 -0.028750  0.173788

In [ ]:
my_tauB_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50 entries, 0 to 49
Data columns (total 51 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Trait Code  50 non-null     object 
 1   Sex         47 non-null     float64
 2   SSSF        47 non-null     float64
 3   ANS         47 non-null     float64
 4   INA         47 non-null     float64
 5   IOB         47 non-null     float64
 6   NAS         47 non-null     float64
 7   NAW         47 non-null     float64
 8   NBC         47 non-null     float64
 9   NBS         47 non-null     float64
 10  NFS         47 non-null     float64
 11  OBS         47 non-null     float64
 12  TPS         47 non-null     float64
 13  INCA        0 non-null      float64
 14  BREG        0 non-null      float64
 15  PALT        47 non-null     float64
 16  METO        47 non-null     float64
 17  MANT        47 non-null     float64
 18  SON         47 non-null     float64
 19  SOF         47 non-null     flo

#### tau B p-value for loop

In [ ]:
# Kendall's tau b p-values
# NA values were dropped from the cnm_tauB df (only 11 individuals in this analysis)

trait_names = ['Trait Code']
trait_names.extend(cnm_tauB.columns)
my_tauB_df3 = pd.DataFrame(columns = trait_names)

for i in range(len(cnm_tauB.columns)):
    temp_df = pd.DataFrame()
    temp_dic = []
    temp_dic.append(cnm_tauB.columns[i])

    for j in range(len(cnm_tauB.columns)):
        tau, p_value = stats.kendalltau(cnm_tauB[cnm_tauB.columns[i]], cnm_tauB[cnm_tauB.columns[j]])
        print(c,'\t', tau, '\t', p_value)
        temp_dic.append(p_value)

    print(len(my_tauB_df.columns), '\t', len(temp_dic))
    my_tauB_df3.loc[i] = temp_dic#[c, tau, p_value]

PHAR 	 1.0 	 7.074463098970675e-10
PHAR 	 -0.02749806661015982 	 0.8639459980780579
PHAR 	 0.01703208390548536 	 0.9124937972451566
PHAR 	 0.0915507056838374 	 0.538684567676416
PHAR 	 -0.13562205717103817 	 0.3793098072587159
PHAR 	 -0.1725123787127004 	 0.28197106580647047
PHAR 	 0.26086956521739135 	 0.10131105093820252
PHAR 	 0.08318015020356473 	 0.5830650177135555
PHAR 	 0.3259710630759294 	 0.0367841744519069
PHAR 	 -0.3935623258746016 	 0.009375343000875485
PHAR 	 -0.14264222271862334 	 0.3626425790575063
PHAR 	 -0.09619325738134539 	 0.5200146809141549
PHAR 	 nan 	 nan
PHAR 	 nan 	 nan
PHAR 	 -0.20755395563871767 	 0.20073916039155448
PHAR 	 0.13530201829834768 	 0.4042484947394712
PHAR 	 0.25425818448750476 	 0.117032270472401
PHAR 	 0.203804347826087 	 0.19638082349986263
PHAR 	 -0.12289192304282605 	 0.42743208750953476
PHAR 	 -0.2545887739877383 	 0.09886741948203089
PHAR 	 -0.0493849928029914 	 0.755621721223902
PHAR 	 0.19781414201873612 	 0.22268927373703729
PHAR 	 -0.0

In [ ]:
my_tauB_df3.head()

Trait Code           Sex          SSSF           ANS           INA           IOB       NAS       NAW       NBC       NBS       NFS       OBS       TPS  INCA  BREG      PALT      METO      MANT       SON       SOF       IFS       ZFF       CCO       FOI       FSI       PTB       TYM  AUD        MT        NO       PZT        ZS       LBM      LBLa        PF        MF       CRB       EPB       FTA       PNB       AST       OMB       HYP       APF        FF       MHB       MEN       MIF       CIV      MFLo      PHAR
0        Sex  7.074463e-10  8.639460e-01  9.124938e-01  5.386846e-01  3.793098e-01  0.281971  0.101311  0.583065  0.036784  0.009375  0.362643  0.520015   NaN   NaN  0.200739  0.404248  0.117032  0.196381  0.427432  0.098867  0.755622  0.222689  0.793717  0.193020  0.387269  0.902837  NaN  0.653626  0.124266  0.523989  0.522593  0.567764  0.307604  0.508108  0.019656  0.703784  0.703784  0.404248  0.939833  0.032879  0.137751  0.972076  0.438239  0.357804  0.290049  0.960687  0.071203  0.112311  0.655554  0.363406
1       SSSF  8.639460e-01  2.992515e-10  7.329247e-01  5.269496e-01  8.129269e-01  0.155988  1.000000  0.984446  0.195925  0.425772  0.949377  0.605533   NaN   NaN  0.470057  0.640795  0.804843  0.566876  0.256769  0.952822  0.309153  0.745304  0.503680  0.022747  0.456266  0.567083  NaN  0.921201  0.783515  0.796958  0.306833  0.011378  0.676338  0.633123  0.143519  0.330879  0.330879  0.640795  0.809336  0.406371  0.452519  0.382789  0.017746  0.004458  0.567083  0.270033  0.446503  0.515891  0.651037  0.516897
2        ANS  9.124938e-01  7.329247e-01  1.461728e-11  2.204413e-02  1.484195e-01  0.250120  0.564749  0.647594  0.878509  0.659036  0.209025  0.001466   NaN   NaN  0.108718  0.625163  0.044197  0.028146  0.009863  0.156707  0.330960  0.504997  0.483812  0.416048  0.427999  0.954384  NaN  0.069959  0.780383  0.801510  0.505579  0.527987  0.284202  0.987313  0.220625  0.014547  0.476046  0.106917  0.051736  0.907719  0.270806  0.805695  0.808720  0.084807  0.567305  0.309482  0.588532  0.561738  0.634399  0.706350
3        INA  5.386846e-01  5.269496e-01  2.204413e-02  2.694645e-13  9.120932e-01  0.451841  0.262613  0.859265  0.704286  0.540817  0.213445  0.391793   NaN   NaN  0.893505  0.092974  0.551296  0.126974  0.195386  0.077332  0.792495  0.466783  0.946678  0.902378  0.840013  0.869806  NaN  0.081109  0.505540  0.218557  0.381957  0.546650  0.596304  0.261086  0.293807  0.158519  0.263426  0.225026  0.084894  0.193274  0.542552  0.106760  0.634151  0.258791  0.675314  0.659034  0.722464  0.842326  0.210516  0.003611
4        IOB  3.793098e-01  8.129269e-01  1.484195e-01  9.120932e-01  9.431898e-12  0.361134  0.021182  0.037802  0.023348  0.585502  0.647587  0.032830   NaN   NaN  0.000409  0.269606  0.015030  0.568065  0.465074  0.348732  0.104433  0.863800  0.630132  0.283538  0.781872  0.119979  NaN  0.204932  0.703600  0.155563  0.141114  0.679350  0.397838  0.511728  0.013840  0.211176  0.499560  0.665750  0.180940  0.886798  0.037671  0.110713  0.020981  0.064206  0.985054  0.180603  0.665870  0.758868  0.842885  0.259921

In [ ]:
my_tauB_df3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50 entries, 0 to 49
Data columns (total 51 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Trait Code  50 non-null     object 
 1   Sex         47 non-null     float64
 2   SSSF        47 non-null     float64
 3   ANS         47 non-null     float64
 4   INA         47 non-null     float64
 5   IOB         47 non-null     float64
 6   NAS         47 non-null     float64
 7   NAW         47 non-null     float64
 8   NBC         47 non-null     float64
 9   NBS         47 non-null     float64
 10  NFS         47 non-null     float64
 11  OBS         47 non-null     float64
 12  TPS         47 non-null     float64
 13  INCA        0 non-null      float64
 14  BREG        0 non-null      float64
 15  PALT        47 non-null     float64
 16  METO        47 non-null     float64
 17  MANT        47 non-null     float64
 18  SON         47 non-null     float64
 19  SOF         47 non-null     flo

#### Output data file:

In [ ]:
#my_tauB_df.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_intertrait_correlations_tauB_unnamed.xlsx', index=False)

In [ ]:
my_tauB_df.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_tauB_intertrait_correlations.xlsx', index=False)

In [ ]:
my_tauB_df3.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_tauB_p-values_intertrait_correlations.xlsx', index=False)

##### Old code:

In [ ]:
'''
# Kendall's tau b

#len(cnm_VarsOnly[cnm_VarsOnly.columns[3]])
column_names = ('Trait Code', 'tau B statistic', 'p-value')
my_tauB_df = pd.DataFrame(columns = column_names)
for c in cnm_VarsOnly.columns:
    temp_df = pd.DataFrame(cnm_VarsOnly[c])                        # initialize a new df using one column [c] of the odont_Jikei_VarsOnly_shapiro at a time
    temp_df.dropna(subset=[c], inplace=True)
    if len(temp_df) <3:                                                             # Shapiro-Wilks cannot run with 3 or less inputs
      continue
    tau, p_value = stats.kendalltau(cnm_VarsOnly['Sex'], temp_df[c])
    print(c,'\t', tau, '\t', p_value)
    #df3 = {'Trait Code': [c], 'tau B statistic': tau[0], 'p-value': p_value[1]}
    my_tauB_df.loc[c] = [c, tau, p_value]
    '''

In [ ]:
'''
# Kendall's tau b

#len(males[males.columns[3]])
column_names = ('Trait Code', 'tau B statistic', 'p-value')
my_tauB_df = pd.DataFrame(columns = column_names)
for c in males.columns:
    temp_df_males = pd.DataFrame(males[c])                                                # initialize a new df using one column [c] of the males df one at a time
    temp_df_males.dropna(subset=[c], inplace=True)
    temp_df_females = pd.DataFrame(females[c])                                                # initialize a new df using one column [c] of the males df one at a time
    temp_df_females.dropna(subset=[c], inplace=True)
    if len(temp_df_males) <3 or len(temp_df_females) <3:                                                             # Levene cannot run with 3 or less inputs
      continue
    tau, p_value = stats.kendalltau(temp_df_females[c], temp_df_males[c])
    print(c,'\t', tau, '\t', p_value)
    #df3 = {'Trait Code': [c], 'tau B statistic': tau[0], 'p-value': p_value[1]}
    my_tauB_df.loc[c] = [c, tau[0], p_value[1]]
    '''

# **Missing data**
Using the cnm_collapsed dataframe, which has had the left and right sides collapsed

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4, SSSF, ANS....

In [ ]:
cnm_cleaned2.head()

In [ ]:
cnm_VarsOnly2 = pd.DataFrame(cnm_cleaned2.iloc[:, 8:])

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SSSF, ANS....

In [ ]:
cnm_VarsOnly2.head()

In [ ]:
percent_missing = cnm_VarsOnly2.isnull().sum().sum()*100 / (len(cnm_VarsOnly2) * len(cnm_VarsOnly2.columns))
percent_missing

5.112560488112771

* The cranial nonmetrics (cnm) dataset has 5.11% missing data.
* For comparison: the odontometrics dataset has 29.51% missing data.

### k-NN imputation

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os

In [ ]:
cnm_kNN_imp = pd.DataFrame(cnm_VarsOnly2)

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SSSF, ANS....

In [ ]:
cnm_kNN_imp.head()

In [ ]:
from sklearn.impute import KNNImputer
imputer = KNNImputer(n_neighbors=5)
imputer.fit(cnm_kNN_imp)
cnm_kNN_imp = imputer.transform(cnm_kNN_imp)          # replace missing values in numeric columns with the average for each column

In [ ]:
print(cnm_kNN_imp)

In [ ]:
cnm_kNN_imp = pd.DataFrame(cnm_kNN_imp)

In [ ]:
# save column names from cnm_VarsOnly2 df

column_names = cnm_VarsOnly2.columns
print(column_names)

Index(['SSSF', 'ANS', 'INA', 'IOB', 'NAS', 'NAW', 'NBC', 'NBS', 'NFS', 'OBS', 'TPS', 'INCA', 'BREG', 'PALT', 'METO', 'MANT', 'SON', 'SOF', 'IFS', 'ZFF', 'CCO', 'FOI', 'FSI', 'PTB', 'TYM', 'AUD', 'MT', 'NO', 'PZT', 'ZS', 'LBM', 'LBLa', 'PF', 'MF', 'CRB', 'EPB', 'FTA', 'PNB', 'AST', 'OMB', 'HYP', 'APF', 'FF', 'MHB', 'MEN', 'MIF', 'CIV', 'MFLo', 'PHAR'], dtype='object')


In [ ]:
# assign column names

cnm_kNN_imp.columns = column_names

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SSSF, ANS....

In [ ]:
cnm_kNN_imp.head()

In [ ]:
cnm_kNN_imp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 0 to 96
Data columns (total 49 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   SSSF    97 non-null     float64
 1   ANS     97 non-null     float64
 2   INA     97 non-null     float64
 3   IOB     97 non-null     float64
 4   NAS     97 non-null     float64
 5   NAW     97 non-null     float64
 6   NBC     97 non-null     float64
 7   NBS     97 non-null     float64
 8   NFS     97 non-null     float64
 9   OBS     97 non-null     float64
 10  TPS     97 non-null     float64
 11  INCA    97 non-null     float64
 12  BREG    97 non-null     float64
 13  PALT    97 non-null     float64
 14  METO    97 non-null     float64
 15  MANT    97 non-null     float64
 16  SON     97 non-null     float64
 17  SOF     97 non-null     float64
 18  IFS     97 non-null     float64
 19  ZFF     97 non-null     float64
 20  CCO     97 non-null     float64
 21  FOI     97 non-null     float64
 22  FSI 

##### Add demographics

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4, SSSF, ANS....

In [ ]:
cnm_cleaned2.head()

The output for the following two code cells were removed to protect PII. The outputs contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4

In [ ]:
selected_columns = cnm_cleaned2[['SkelID', 'Collection', 'Sex', 'Age', 'Population', 'Population2', 'Population3', 'Population4']]
dems2 = selected_columns.copy()
dems2.head()

In [ ]:
dems2.tail()

In [ ]:
cnm_kNN_imp = pd.merge(dems2, cnm_kNN_imp, left_index=True, right_index=True)

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4, SSSF, ANS....

In [ ]:
cnm_kNN_imp.head()

In [ ]:
cnm_kNN_imp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 0 to 96
Data columns (total 57 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   SkelID       97 non-null     int64  
 1   Collection   97 non-null     object 
 2   Sex          97 non-null     object 
 3   Age          97 non-null     int64  
 4   Population   97 non-null     object 
 5   Population2  19 non-null     object 
 6   Population3  10 non-null     object 
 7   Population4  1 non-null      object 
 8   SSSF         97 non-null     float64
 9   ANS          97 non-null     float64
 10  INA          97 non-null     float64
 11  IOB          97 non-null     float64
 12  NAS          97 non-null     float64
 13  NAW          97 non-null     float64
 14  NBC          97 non-null     float64
 15  NBS          97 non-null     float64
 16  NFS          97 non-null     float64
 17  OBS          97 non-null     float64
 18  TPS          97 non-null     float64
 19  INCA      

##### Output data file: k-NN imputation

In [ ]:
cnm_kNN_imp.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_kNN_imp.xlsx', index=False)

### Iterative imputation

In [ ]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import os

In [ ]:
cnm_iterative_imp = pd.DataFrame(cnm_VarsOnly2)

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SSSF, ANS....

In [ ]:
cnm_iterative_imp.head()

In [ ]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer
imputer = IterativeImputer(max_iter=200, random_state=0)
imputer.fit(cnm_iterative_imp)
cnm_iterative_imp = imputer.transform(cnm_iterative_imp)

In [ ]:
print(cnm_iterative_imp)

In [ ]:
cnm_iterative_imp = pd.DataFrame(cnm_iterative_imp)

In [ ]:
# save column names from cnm_VarsOnly2 df

column_names = cnm_VarsOnly2.columns
print(column_names)

Index(['SSSF', 'ANS', 'INA', 'IOB', 'NAS', 'NAW', 'NBC', 'NBS', 'NFS', 'OBS', 'TPS', 'INCA', 'BREG', 'PALT', 'METO', 'MANT', 'SON', 'SOF', 'IFS', 'ZFF', 'CCO', 'FOI', 'FSI', 'PTB', 'TYM', 'AUD', 'MT', 'NO', 'PZT', 'ZS', 'LBM', 'LBLa', 'PF', 'MF', 'CRB', 'EPB', 'FTA', 'PNB', 'AST', 'OMB', 'HYP', 'APF', 'FF', 'MHB', 'MEN', 'MIF', 'CIV', 'MFLo', 'PHAR'], dtype='object')


In [ ]:
# assign column names

cnm_iterative_imp.columns = column_names

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SSSF, ANS....

In [ ]:
cnm_iterative_imp.head()

In [ ]:
cnm_iterative_imp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 0 to 96
Data columns (total 49 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   SSSF    97 non-null     float64
 1   ANS     97 non-null     float64
 2   INA     97 non-null     float64
 3   IOB     97 non-null     float64
 4   NAS     97 non-null     float64
 5   NAW     97 non-null     float64
 6   NBC     97 non-null     float64
 7   NBS     97 non-null     float64
 8   NFS     97 non-null     float64
 9   OBS     97 non-null     float64
 10  TPS     97 non-null     float64
 11  INCA    97 non-null     float64
 12  BREG    97 non-null     float64
 13  PALT    97 non-null     float64
 14  METO    97 non-null     float64
 15  MANT    97 non-null     float64
 16  SON     97 non-null     float64
 17  SOF     97 non-null     float64
 18  IFS     97 non-null     float64
 19  ZFF     97 non-null     float64
 20  CCO     97 non-null     float64
 21  FOI     97 non-null     float64
 22  FSI 

##### Add demographics

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4, SSSF, ANS....

In [ ]:
cnm_cleaned2.head()

The output for the following two code cells were removed to protect PII. The outputs contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4

In [ ]:
#selected_columns = cnm_collapsed[['SkelID', 'Collection', 'Sex', 'Age', 'Population', 'Population2', 'Population3', 'Population4']]
#dems2 = selected_columns.copy()
dems2.head()

In [ ]:
dems2.tail()

In [ ]:
cnm_iterative_imp = pd.merge(dems2, cnm_iterative_imp, left_index=True, right_index=True)

The output for the following code cell was removed to protect PII. The output contained the following columns:
* SkelID, Collection, Sex, Age, Population, Population2, Population3, Population4, SSSF, ANS....

In [ ]:
cnm_iterative_imp.head()

In [ ]:
cnm_iterative_imp.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 97 entries, 0 to 96
Data columns (total 57 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   SkelID       97 non-null     int64  
 1   Collection   97 non-null     object 
 2   Sex          97 non-null     object 
 3   Age          97 non-null     int64  
 4   Population   97 non-null     object 
 5   Population2  19 non-null     object 
 6   Population3  10 non-null     object 
 7   Population4  1 non-null      object 
 8   SSSF         97 non-null     float64
 9   ANS          97 non-null     float64
 10  INA          97 non-null     float64
 11  IOB          97 non-null     float64
 12  NAS          97 non-null     float64
 13  NAW          97 non-null     float64
 14  NBC          97 non-null     float64
 15  NBS          97 non-null     float64
 16  NFS          97 non-null     float64
 17  OBS          97 non-null     float64
 18  TPS          97 non-null     float64
 19  INCA      

##### Output data file: iterative imputation

In [ ]:
cnm_iterative_imp.to_excel('/drive/My Drive/Colab Notebooks/Pre-statistical treatments/2_EDA/cnm_iterative_imp.xlsx', index=False)

# Output data files:
*   *cnm_cleaned2.xlsx*: This file includes the cranial nonmetric and macromorphoscopic data and demographics after traits with > 50% missing data were removed.
*   *cnm_collapsed.xlsx*: This file contains the cnm data with left and right sides collapsed (prior to having traits with > 50% data missing removed). The traits have been renamed and traits with low IO agreement have already been removed.




##### **Sex: Need to finish outputting these**
##### **Sexual dimorphism**
*   *cnm_SD_chi_square.xlsx*: This file contains the chi-square test for independence results for sexual dimorphism.
##### **Intertrait correlations**
*   *cnm_tauB_intertrait_correlations.xlsx*
*   *cnm_tauB_p-values_intertrait_correlations.xlsx*
*   These^ files contain the Kendall's tau B results and associated p-values.
##### **Missing data**
*   *cnm_kNN_imp.xlsx*: This file contains the cnm dataset with missing values imputed using the k-NN imputation method (k = 5). The measurement variables were rounded to 2 decimal places in excel.
*   *cnm_iterative_imp.xlsx*: This file contains the cnm dataset with missing values imputed using the iterative imputation method (max 200 iterations). The measurement variables were rounded to 2 decimal places in Excel.
